src/devices/src/virtio/block.rs

```diff
@@ -300,27 +294,81 @@ impl Request {
     }
 }
 
-/// Handler that drives the execution of the Block devices
-pub struct BlockEpollHandler {
-    queues: Vec<Queue>,
-    mem: GuestMemory,
+/// Virtio device for exposing block level read/write operations on a host file.
+pub struct Block {
     disk_image: File,
     disk_nsectors: u64,
-    interrupt_status: Arc<AtomicUsize>,
-    interrupt_evt: EventFd,
-    queue_evt: EventFd,
+    avail_features: u64,
+    acked_features: u64,
+    config_space: Vec<u8>,
     rate_limiter: RateLimiter,
+    queues: Option<Vec<Queue>>,
+    mem: Option<GuestMemory>,
+    interrupt_status: Option<Arc<AtomicUsize>>,
+    interrupt_evt: Option<EventFd>,
+    queue_evt: Option<EventFd>,
     disk_image_id: Vec<u8>,
 }
 
-impl BlockEpollHandler {
+pub fn build_config_space(disk_size: u64) -> Vec<u8> {
+    // We only support disk size, which uses the first two words of the configuration space.
+    // If the image is not a multiple of the sector size, the tail bits are not exposed.
+    // The config space is little endian.
+    let mut config = Vec::with_capacity(CONFIG_SPACE_SIZE);
+    let num_sectors = disk_size >> SECTOR_SHIFT;
+    for i in 0..8 {
+        config.push((num_sectors >> (8 * i)) as u8);
+    }
+    config
+}
+
+impl Block {
+    /// Create a new virtio block device that operates on the given file.
+    ///
+    /// The given file must be seekable and sizable.
+    pub fn new(
+        mut disk_image: File,
+        is_disk_read_only: bool,
+        rate_limiter: RateLimiter,
+    ) -> io::Result<Block> {
+        let disk_size = disk_image.seek(SeekFrom::End(0))? as u64;
+        if disk_size % SECTOR_SIZE != 0 {
+            warn!(
+                "Disk size {} is not a multiple of sector size {}; \
+                 the remainder will not be visible to the guest.",
+                disk_size, SECTOR_SIZE
+            );
+        }
+
+        let mut avail_features = (1u64 << VIRTIO_F_VERSION_1) | (1u64 << VIRTIO_BLK_F_FLUSH);
+
+        if is_disk_read_only {
+            avail_features |= 1u64 << VIRTIO_BLK_F_RO;
+        };
+
+        Ok(Block {
+            disk_image_id: build_disk_image_id(&disk_image),
+            disk_image: disk_image,
+            disk_nsectors: disk_size / SECTOR_SIZE,
+            avail_features,
+            acked_features: 0u64,
+            config_space: build_config_space(disk_size),
+            rate_limiter,
+            interrupt_status: None,
+            interrupt_evt: None,
+            queue_evt: None,
+            queues: None,
+            mem: None,
+        })
+    }
+
     fn process_queue(&mut self, queue_index: usize) -> bool {
-        let queue = &mut self.queues[queue_index];
+        let queues = &mut self.queues.as_mut().unwrap();
```

Comment:
Removing the `Option` from `queues` and event_fds will make this code easier and also avoid these dangerous `unwrap()`s.